refinery/file_store/tasks.py

```diff
@@ -20,11 +21,129 @@
 from data_set_manager.search_indexes import NodeIndex
 
 from .models import FileStoreItem, _mkdir, get_temp_dir, parse_s3_url
+from .utils import S3MediaStorage, SymlinkedFileSystemStorage, _delete_file
 
 logger = celery.utils.log.get_task_logger(__name__)
 logger.setLevel(celery.utils.LOG_LEVELS['DEBUG'])
 
 
+class ProgressPercentage(object):
+    """Callable for progress monitoring of S3 transfers"""
+
+    def __init__(self, filename, task_id):
+        self._filename = filename
+        self._task_id = task_id
+        self._size = os.path.getsize(filename)
+        self._seen_so_far = 0
+        self._lock = threading.Lock()
+
+    def __call__(self, bytes_amount):
+        # to simplify we'll assume this is hooked up to a single filename
+        with self._lock:
+            self._seen_so_far += bytes_amount
+            percentage = self._seen_so_far * 100. / self._size
+            import_file.update_state(
+                self._task_id, state='PROGRESS',
+                meta={
+                    'percent_done': '{:.0f}'.format(percentage),
+                    'current': self._seen_so_far,
+                    'total': self._size
+                }
+            )
+
+
+def _copy_file_obj(source, destination):
+    """Copy file object in chunks"""
+    chunk_size = 10 * 1024 * 1024  # 10MB
+    source_size = os.path.getsize(source.name)
+    for chunk in iter(lambda: source.read(chunk_size), ''):
+        destination.write(chunk)
+        destination_size = os.path.getsize(destination.name)
+        percentage = destination_size * 100. / source_size
+        import_file.update_state(
+            state='PROGRESS',
+            meta={
+                'percent_done': '{:.0f}'.format(percentage),
+                'current': destination_size,
+                'total': source_size
+            }
+        )
+
+
+def _copy_file(source_path, destination_path, delete_source=False):
+    """Copy file from one absolute file system path to another"""
+    _mkdir(os.path.dirname(destination_path))
+
+    try:
+        with open(source_path, 'rb') as source, \
+                open(destination_path, 'wb') as destination:
+            _copy_file_obj(source, destination)
+    except IOError as exc:
+        _delete_file(destination_path)
+        raise RuntimeError("Failed to copy '{}' to '{}': {}".format(
+            source_path, destination_path, exc
+        ))
+
+    if delete_source:
+        _delete_file(source_path)
+
+
+def _symlink_file(source_path, link_path):
+    _mkdir(os.path.dirname(link_path))
+    try:
+        os.symlink(source_path, link_path)
+    except OSError as exc:
+        raise RuntimeError("Error creating symlink '%s': %s", link_path, exc)
```

Comment:
Should the dir created from: `_mkdir(os.path.dirname(link_path))` be removed on error?

Reply:
I don't think it is necessary in this context.